File: services/translation-worker/vantara_worker/layout.py
"""تخطيط العربي داخل الفقاعة: التفاف مقيَّد بمضلّع الفقاعة لا بمستطيل.

الفكرة كما يفعل المخطِّط البشري: يبدأ من أكبر حجم خط، ويلفّ الكلمات في أسطر
كل سطر بعرض الوتر المتاح داخل الفقاعة عند ارتفاعه (من قناع الفقاعة المتآكل)،
ويوازن الأسطر (الأوسط أعرض، الأطراف أضيق)، وينزل حجمًا حتى تدخل الكتلة كلها.

الرسم بـPillow/RAQM (HarfBuzz + FriBidi) — الطبقة المُثبتة في `arabic.py`.
الخط Baloo Bhaijaan 2 (OFL): مصمَّم للعربية، مستدير سميك يشبه خطوط الكوميك.

النص فوق الرسم (بلا فقاعة) يُرسم في صندوق النص الأصلي بلون حبره (أبيض على
داكن، أسود على فاتح) مع حدّ بلون مضاد بسُمك يتبع الحجم، كما في الأصل.
"""

from __future__ import annotations

import contextlib
from dataclasses import dataclass
from functools import lru_cache
from pathlib import Path

import cv2
import numpy as np
from PIL import Image, ImageDraw, ImageFont

from .regions import Region
from .vision.common import Box

FONT_PATH = str(Path(__file__).resolve().parent.parent / "fonts" / "BalooBhaijaan2.ttf")
MIN_SIZE = 12
LINE_SPACING = 1.08  # من ارتفاع الحرف الفعلي لا من ascent+descent (Baloo يبالغ فيهما)


@lru_cache(maxsize=128)
def font(size: int, path: str = FONT_PATH) -> ImageFont.FreeTypeFont:
    f = ImageFont.truetype(path, size, layout_engine=ImageFont.Layout.RAQM)
    with contextlib.suppress(Exception):
        f.set_variation_by_name("SemiBold")
    return f


_measure = ImageDraw.Draw(Image.new("L", (1, 1)))


def text_width(text: str, size: int) -> float:
    return _measure.textlength(text, font=font(size), direction="rtl", language="ar")


def line_metrics(size: int) -> tuple[int, int]:
    """(ارتفاع سطر، إزاحة القاعدة) من صندوق حبر حقيقي لعيّنة عربية عالية ومنخفضة."""
    f = font(size)
    _l, t, _r, b = f.getbbox("لجفقيبطئ", anchor="ls", direction="rtl", language="ar")
    ascent = -t
    descent = b
    return round((ascent + descent) * LINE_SPACING), int(ascent)


@dataclass
class Layout:
    size: int
    lines: list[str]
    positions: list[tuple[float, float]]  # مركز x، قاعدة y لكل سطر
    bounds: Box
    fits: bool


def _chord(mask: np.ndarray, y: int, cx: int) -> tuple[int, int] | None:
    """أطول امتداد أفقي داخل القناع عند الصف y يحتوي cx (أو الأقرب إليه)."""
    if y < 0 or y >= mask.shape[0]:
        return None
    row = mask[y] > 0
    if not row.any():
        return None
    xs = np.where(row)[0]
    # مقاطع متصلة
    breaks = np.where(np.diff(xs) > 1)[0]
    segs = np.split(xs, breaks + 1)
    best = min(segs, key=lambda s: 0 if s[0] <= cx <= s[-1] else min(abs(s[0] - cx), abs(s[-1] - cx)))
    return int(best[0]), int(best[-1])


def _wrap_polygon(words: list[str], size: int, mask: np.ndarray, cx: int, top: int, line_h: int, pad: int) -> tuple[list[str], list[tuple[int, int]]] | None:
    """يلفّ الكلمات سطرًا سطرًا من `top` نزولًا؛ كل سطر بعرض وتر القناع عنده."""
    lines: list[str] = []
    spans: list[tuple[int, int]] = []
    y = top
    i = 0
    while i < len(words):
        rows = [_chord(mask, yy, cx) for yy in (y, y + line_h // 2, y + line_h - 1)]
        if any(r is None for r in rows):
            return None
        left = max(r[0] for r in rows) + pad  # type: ignore[index]
        right = min(r[1] for r in rows) - pad  # type: ignore[index]
        width = right - left
        if width < size:
            return None
        line = words[i]
        if text_width(line, size) > width:
            return None  # كلمة واحدة أطول من الوتر: لا نقطع الكلمة العربية
        i += 1
        while i < len(words) and text_width(f"{line} {words[i]}", size) <= width:
            line = f"{line} {words[i]}"
            i += 1
        lines.append(line)
        spans.append((left, right))
        y += line_h
    return lines, spans


def fit_in_mask(text: str, mask: np.ndarray, anchor: Box, *, max_size: int, min_size: int = MIN_SIZE, pad: int = 4) -> Layout | None:
    """أكبر حجم يُدخل النص في القناع، متمركزًا رأسيًا حول مركز صندوق النص الأصلي."""
    words = text.split()
    if not words:
        return None
    ys, _xs = np.where(mask > 0)
    if len(ys) == 0:
        return None
    cx = int((anchor[0] + anchor[2]) / 2)
    cy_anchor = (anchor[1] + anchor[3]) / 2
    cy_mask = float(ys.mean())
    # المركز: مركز النص الأصلي، لكنه لا يبتعد عن مركز الفقاعة أكثر من الربع
    cy = cy_anchor if abs(cy_anchor - cy_mask) < (ys.max() - ys.min()) * 0.25 else cy_mask
    for size in range(max_size, min_size - 1, -1):
        line_h, ascent = line_metrics(size)
        # نخمّن عدد الأسطر، نلفّ من أعلى الكتلة المتمركزة، ثم نعيد التمركز على العدد الفعلي
        n_guess = 1
        for _ in range(14):
            top = int(cy - n_guess * line_h / 2)
            res = _wrap_polygon(words, size, mask, cx, top, line_h, pad)
            if res is None:
                n_guess += 1
                continue
            lines, spans = res
            if len(lines) != n_guess:
                n_guess = len(lines)
                res2 = _wrap_polygon(words, size, mask, cx, int(cy - n_guess * line_h / 2), line_h, pad)
                if res2 is None or len(res2[0]) != n_guess:
                    n_guess += 1
                    continue
                lines, spans = res2
                top = int(cy - n_guess * line_h / 2)
            positions = []
            x1 = 10**9
            x2 = -1
            for k, (l, r) in enumerate(spans):
                w = text_width(lines[k], size)
                mid = (l + r) / 2
                positions.append((mid, top + k * line_h + ascent))
                x1 = min(x1, int(mid - w / 2))
                x2 = max(x2, int(mid + w / 2))
            return Layout(size, lines, positions, (x1, top, x2, top + len(lines) * line_h), True)
    return None


def fit_in_box(text: str, box: Box, *, max_size: int, image_size: tuple[int, int], min_size: int = MIN_SIZE, grow: float = 1.2) -> Layout | None:
    """للنص فوق الرسم: مستطيل النص الأصلي موسَّعًا قليلًا، ولا يخرج من الصورة."""
    words = text.split()
    if not words:
        return None
    W, H = image_size
    x1, y1, x2, y2 = box
    cx, cy = (x1 + x2) / 2, (y1 + y2) / 2
    w = (x2 - x1) * grow
    h = (y2 - y1) * grow
    margin = 8
    w = min(w, 2 * (cx - margin), 2 * (W - margin - cx))
    h = min(h, 2 * (cy - margin), 2 * (H - margin - cy))
    for size in range(max_size, min_size - 1, -1):
        line_h, ascent = line_metrics(size)
        lines: list[str] = []
        cur = ""
        ok = True
        for word in words:
            cand = f"{cur} {word}".strip()
            if not cur or text_width(cand, size) <= w:
                cur = cand
            else:
                if text_width(cur, size) > w:
                    ok = False
                lines.append(cur)
                cur = word
        if cur:
            lines.append(cur)
        if not ok or any(text_width(line, size) > w for line in lines) or len(lines) * line_h > h:
            continue
        top = cy - len(lines) * line_h / 2
        positions = [(cx, top + k * line_h + ascent) for k in range(len(lines))]
        widest = max(text_width(line, size) for line in lines)
        return Layout(size, lines, positions, (int(cx - widest / 2), int(top), int(cx + widest / 2), int(top + len(lines) * line_h)), True)
    return None


def glyph_height_of(region: Region) -> int:
    x1, y1, x2, y2 = region.box
    sub = region.glyph[y1:y2, x1:x2] > 0
    rows = sub.sum(axis=1) > 0
    runs: list[int] = []
    n = 0
    for v in rows:
        if v:
            n += 1
        elif n:
            runs.append(n)
            n = 0
    if n:
        runs.append(n)
    return int(np.median(runs)) if runs else max(12, (y2 - y1) // 3)


def layout_region(region: Region, text: str, inner_mask: np.ndarray | None, image_size: tuple[int, int]) -> Layout | None:
    """يختار المساحة الصحيحة للمنطقة ويعيد تخطيطًا أو None إن لم يدخل النص بحجم مقروء."""
    gh = glyph_height_of(region)
    # حجم الخط: يتبع ارتفاع الحرف الأصلي. العربي بلا حروف كبيرة فيبدو أصغر عند نفس
    # الحجم، لذا داخل الفقاعة نسمح حتى ×1.7 (المضلّع يوقفه)، وفوق الرسم ×1.15 (الصندوق يوقفه)
    if inner_mask is not None:
        max_size = int(max(MIN_SIZE + 4, min(gh * 2.0, 110)))
        lay = fit_in_mask(text, inner_mask, region.box, max_size=max_size)
        if lay is not None:
            return lay
    max_size = int(max(MIN_SIZE + 4, min(gh * 1.15, 96)))
    return fit_in_box(text, region.box, max_size=max_size, image_size=image_size)


def draw_layout(image: Image.Image, layout: Layout, *, ink_light: bool, on_art: bool) -> None:
    draw = ImageDraw.Draw(image)
    f = font(layout.size)
    fill = (255, 255, 255) if ink_light else (16, 16, 16)
    stroke = (16, 16, 16) if ink_light else (255, 255, 255)
    stroke_w = max(2, layout.size // 9) if on_art else 0
    for line, (cx, base) in zip(layout.lines, layout.positions, strict=True):
        draw.text(
            (cx, base),
            line,
            font=f,
            fill=fill,
            anchor="ms",
            direction="rtl",
            language="ar",
            stroke_width=stroke_w,
            stroke_fill=stroke if stroke_w else None,
        )


def inner_mask_for(region: Region, rgb: np.ndarray, siblings: list[Region] | None = None) -> np.ndarray | None:
    """داخل الفقاعة المتآكل: هنا وحده يُرسم العربي. فقاعة تحمل جملتين تُقتسم:
    نصيب كل منطقة هو داخل الفقاعة بعيدًا عن صناديق شقيقاتها."""
    mask = None
    if region.bubble is not None:
        mask = region.bubble.mask
    elif region.bubble_box is not None:
        from .regions import flat_box_mask

        mask = flat_box_mask(rgb, region.bubble_box, region.box)
    if mask is None:
        return None
    gh = glyph_height_of(region)
    erode = max(6, int(gh * 0.45))
    inner = cv2.erode(mask, np.ones((erode * 2 + 1, erode * 2 + 1), np.uint8))
    for sib in siblings or []:
        sx1, sy1, sx2, sy2 = sib.box
        gap = max(4, gh // 2)
        # الحدّ بين الجملتين: منتصف المسافة بينهما، لا صندوق الشقيقة وحده
        _x1, y1, _x2, y2 = region.box
        if sy1 >= y2:  # الشقيقة تحت
            cut = (y2 + sy1) // 2
            inner[cut - gap // 2 :, :] = 0
        elif sy2 <= y1:  # الشقيقة فوق
            cut = (sy2 + y1) // 2
            inner[: cut + gap // 2, :] = 0
        else:
            inner[max(0, sy1 - gap) : sy2 + gap, max(0, sx1 - gap) : sx2 + gap] = 0
    return inner

File: android/app/src/main/kotlin/com/vantara/plugins/translation/Layout.kt
package com.vantara.plugins.translation

import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Typeface

/** تخطيط نص عربي: الحجم والأسطر ومركز كل سطر وقاعدته وحدود الكتلة. */
class TextLayout(val size: Float, val lines: List<String>, val centers: List<FloatArray>, val bounds: Box, val lineBounds: List<Box>)

/**
 * تخطيط العربي داخل الفقاعة — ترجمة `vantara_worker/layout.py`:
 * التفاف مقيَّد بوتر الفقاعة عند كل سطر، تمركز حول مركز النص الأصلي، أكبر
 * حجم يدخل. القياس والرسم بـ`Paint` أندرويد: تشكيل العربي والاتجاه من النظام
 * نفسه (HarfBuzz/ICU داخل Minikin)، بخط Baloo Bhaijaan 2 المرفق.
 */
class ArabicLayout(private val typeface: Typeface) {
    private val paint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        typeface = this@ArabicLayout.typeface
        textAlign = Paint.Align.CENTER
        isFakeBoldText = false
    }
    private val minSize = 12f
    private val lineSpacing = 1.08f

    private fun width(text: String, size: Float): Float {
        paint.textSize = size
        return paint.measureText(text)
    }

    /** ارتفاع السطر وصعود القاعدة من صندوق حبر عيّنة عربية عالية ومنخفضة. */
    private fun metrics(size: Float): Pair<Float, Float> {
        paint.textSize = size
        val r = android.graphics.Rect()
        paint.getTextBounds("لجفقيبطئ", 0, 8, r)
        val ascent = -r.top.toFloat()
        val descent = r.bottom.toFloat()
        return (ascent + descent) * lineSpacing to ascent
    }

    private fun wrapPolygon(words: List<String>, size: Float, mask: ByteMask, cx: Int, top: Int, lineH: Float, pad: Int): Pair<List<String>, List<IntArray>>? {
        val lines = ArrayList<String>()
        val spans = ArrayList<IntArray>()
        var y = top.toFloat()
        var i = 0
        while (i < words.size) {
            val rows = listOf(y.toInt(), (y + lineH / 2).toInt(), (y + lineH - 1).toInt()).map { mask.chord(it, cx) }
            if (rows.any { it == null }) return null
            val left = rows.maxOf { it!![0] } + pad
            val right = rows.minOf { it!![1] } - pad
            val w = right - left
            if (w < size) return null
            var line = words[i]
            if (width(line, size) > w) return null
            i++
            while (i < words.size && width("$line ${words[i]}", size) <= w) { line = "$line ${words[i]}"; i++ }
            lines.add(line)
            spans.add(intArrayOf(left, right))
            y += lineH
        }
        return lines to spans
    }

    fun fitInMask(text: String, mask: ByteMask, anchor: Box, maxSize: Float, pad: Int = 4): TextLayout? {
        val words = text.trim().split(Regex("\\s+")).filter { it.isNotEmpty() }
        if (words.isEmpty()) return null
        val b = mask.bounds() ?: return null
        var sumY = 0L; var n = 0L
        for (y in b[1] until b[3]) for (x in b[0] until b[2]) if (mask[x, y].toInt() != 0) { sumY += y; n++ }
        if (n == 0L) return null
        val cx = (anchor.x1 + anchor.x2) / 2
        val cyAnchor = (anchor.y1 + anchor.y2) / 2f
        val cyMask = sumY.toFloat() / n
        val cy = if (Math.abs(cyAnchor - cyMask) < (b[3] - b[1]) * 0.25f) cyAnchor else cyMask
        var size = maxSize
        while (size >= minSize) {
            val (lineH, ascent) = metrics(size)
            var guess = 1
            var tries = 0
            while (tries++ < 14) {
                var top = (cy - guess * lineH / 2).toInt()
                var res = wrapPolygon(words, size, mask, cx, top, lineH, pad)
                if (res == null) { guess++; continue }
                if (res.first.size != guess) {
                    guess = res.first.size
                    top = (cy - guess * lineH / 2).toInt()
                    val res2 = wrapPolygon(words, size, mask, cx, top, lineH, pad)
                    if (res2 == null || res2.first.size != guess) { guess++; continue }
                    res = res2
                }
                val (lines, spans) = res
                val centers = ArrayList<FloatArray>()
                val lb = ArrayList<Box>()
                var x1 = Int.MAX_VALUE; var x2 = -1
                for (k in lines.indices) {
                    val w = width(lines[k], size)
                    val mid = (spans[k][0] + spans[k][1]) / 2f
                    val base = top + k * lineH + ascent
                    centers.add(floatArrayOf(mid, base))
                    val lx1 = (mid - w / 2).toInt(); val lx2 = (mid + w / 2).toInt()
                    lb.add(Box(lx1, (base - ascent).toInt(), lx2, (base - ascent + lineH).toInt()))
                    if (lx1 < x1) x1 = lx1; if (lx2 > x2) x2 = lx2
                }
                return TextLayout(size, lines, centers, Box(x1, top, x2, (top + lines.size * lineH).toInt()), lb)
            }
            size -= 1f
        }
        return null
    }

    fun fitInBox(text: String, box: Box, maxSize: Float, imageW: Int, imageH: Int, grow: Float = 1.2f): TextLayout? {
        val words = text.trim().split(Regex("\\s+")).filter { it.isNotEmpty() }
        if (words.isEmpty()) return null
        val cx = (box.x1 + box.x2) / 2f; val cy = (box.y1 + box.y2) / 2f
        val margin = 8f
        val w = minOf(box.w * grow, 2 * (cx - margin), 2 * (imageW - margin - cx))
        val h = minOf(box.h * grow, 2 * (cy - margin), 2 * (imageH - margin - cy))
        var size = maxSize
        while (size >= minSize) {
            val (lineH, ascent) = metrics(size)
            val lines = ArrayList<String>()
            var cur = ""
            var ok = true
            for (word in words) {
                val cand = if (cur.isEmpty()) word else "$cur $word"
                if (cur.isEmpty() || width(cand, size) <= w) cur = cand
                else { if (width(cur, size) > w) ok = false; lines.add(cur); cur = word }
            }
            if (cur.isNotEmpty()) lines.add(cur)
            if (!ok || lines.any { width(it, size) > w } || lines.size * lineH > h) { size -= 1f; continue }
            val top = cy - lines.size * lineH / 2
            val centers = lines.indices.map { floatArrayOf(cx, top + it * lineH + ascent) }
            val widest = lines.maxOf { width(it, size) }
            val lb = lines.indices.map { k -> Box((cx - width(lines[k], size) / 2).toInt(), (top + k * lineH).toInt(), (cx + width(lines[k], size) / 2).toInt(), (top + (k + 1) * lineH).toInt()) }
            return TextLayout(size, lines, centers, Box((cx - widest / 2).toInt(), top.toInt(), (cx + widest / 2).toInt(), (top + lines.size * lineH).toInt()), lb)
        }
        return null
    }

    /** داخل الفقاعة المتآكل، مقتسمًا مع الشقيقات. */
    fun innerMaskFor(img: RgbImage, region: Region, siblings: List<Region>?): ByteMask? {
        val mask = bubbleMaskFor(img, region) ?: return null
        return innerOf(mask, region, siblings, maxOf(6, (Cleaner.glyphHeight(region.glyph, region.box) * 0.45).toInt()))
    }

    private fun bubbleMaskFor(img: RgbImage, region: Region): ByteMask? =
        region.bubble?.mask ?: region.bubbleBox?.let { Regions.flatBoxMask(img, it, region.box) }

    private fun innerOf(mask: ByteMask, region: Region, siblings: List<Region>?, erode: Int): ByteMask {
        val gh = Cleaner.glyphHeight(region.glyph, region.box)
        val inner = mask.erode(erode)
        for (sib in siblings ?: emptyList()) {
            val gap = maxOf(4, gh / 2)
            if (sib.box.y1 >= region.box.y2) inner.fillRect(0, (region.box.y2 + sib.box.y1) / 2 - gap / 2, inner.width, inner.height, 0)
            else if (sib.box.y2 <= region.box.y1) inner.fillRect(0, 0, inner.width, (sib.box.y2 + region.box.y1) / 2 + gap / 2, 0)
            else inner.fillRect(sib.box.x1 - gap, sib.box.y1 - gap, sib.box.x2 + gap, sib.box.y2 + gap, 0)
        }
        return inner
    }

    /**
     * فقاعة معروفة: العربي داخل مضلّعها وحده، لا يخرج إلى الرسم أبدًا. إن لم يدخل
     * في الداخل المتآكل يُجرَّب هامش أرفع، ثم الصندوق بشرط أن يقع كله داخل الفقاعة؛
     * وإلا لا تخطيط (تبقى الفقاعة كما هي). النص الحر بلا فقاعة: الصندوق كما كان.
     */
    fun layoutRegion(img: RgbImage, region: Region, text: String, siblings: List<Region>?): TextLayout? {
        val gh = Cleaner.glyphHeight(region.glyph, region.box)
        val boxSize = maxOf(minSize + 4, minOf(gh * 1.15f, 96f))
        val bubble = bubbleMaskFor(img, region)
            ?: return fitInBox(text, region.box, boxSize, img.width, img.height)
        val maxSize = maxOf(minSize + 4, minOf(gh * 2.0f, 110f))
        val erode = maxOf(6, (gh * 0.45).toInt())
        fitInMask(text, innerOf(bubble, region, siblings, erode), region.box, maxSize)?.let { return it }
        val thin = maxOf(3, erode / 2)
        val inner = innerOf(bubble, region, siblings, thin)
        fitInMask(text, inner, region.box, maxSize, pad = 2)?.let { return it }
        return fitInBox(text, region.box, boxSize, img.width, img.height)?.takeIf { insideMask(it, inner) }
    }

    /** كل سطر (بزواياه ومنتصف حوافه) داخل القناع. */
    private fun insideMask(l: TextLayout, mask: ByteMask): Boolean = l.lineBounds.all { b ->
        val xs = intArrayOf(b.x1, (b.x1 + b.x2) / 2, b.x2 - 1)
        val ys = intArrayOf(b.y1, (b.y1 + b.y2) / 2, b.y2 - 1)
        xs.all { x -> ys.all { y -> x in 0 until mask.width && y in 0 until mask.height && mask[x, y].toInt() != 0 } }
    }

    /** يرسم التخطيط على الـBitmap: لون الحبر الأصلي، وحدّ مضاد فوق الرسم. */
    fun draw(canvas: Canvas, layout: TextLayout, inkLight: Boolean, onArt: Boolean) {
        val fill = Paint(paint).apply {
            textSize = layout.size
            color = if (inkLight) 0xFFFFFFFF.toInt() else 0xFF101010.toInt()
            style = Paint.Style.FILL
        }
        val stroke = if (onArt) Paint(paint).apply {
            textSize = layout.size
            color = if (inkLight) 0xFF101010.toInt() else 0xFFFFFFFF.toInt()
            style = Paint.Style.STROKE
            strokeWidth = maxOf(2f, layout.size / 9f) * 2
            strokeJoin = Paint.Join.ROUND
        } else null
        for (k in layout.lines.indices) {
            val (cx, base) = layout.centers[k].let { it[0] to it[1] }
            stroke?.let { canvas.drawText(layout.lines[k], cx, base, it) }
            canvas.drawText(layout.lines[k], cx, base, fill)
        }
    }

    companion object {
        fun bitmapOf(img: RgbImage): Bitmap {
            val bmp = Bitmap.createBitmap(img.width, img.height, Bitmap.Config.ARGB_8888)
            val px = IntArray(img.width * img.height)
            for (i in px.indices) px[i] = (0xff shl 24) or ((img.data[i * 3].toInt() and 0xff) shl 16) or ((img.data[i * 3 + 1].toInt() and 0xff) shl 8) or (img.data[i * 3 + 2].toInt() and 0xff)
            bmp.setPixels(px, 0, img.width, 0, 0, img.width, img.height)
            return bmp
        }

        fun rgbOf(bmp: Bitmap): RgbImage {
            val w = bmp.width; val h = bmp.height
            val px = IntArray(w * h)
            bmp.getPixels(px, 0, w, 0, 0, w, h)
            val out = ByteArray(w * h * 3)
            for (i in px.indices) {
                out[i * 3] = (px[i] shr 16).toByte(); out[i * 3 + 1] = (px[i] shr 8).toByte(); out[i * 3 + 2] = px[i].toByte()
            }
            return RgbImage(w, h, out)
        }
    }
}
